File: camount/src/commonMain/kotlin/io/denison/camount/formatter/FieldRange.kt
package io.denison.camount.formatter

internal data class FieldRange(
  var beginIndex: Int = 0,
  var endIndex: Int = 0,
) {
  val length: Int get() = if (isValid()) endIndex - beginIndex else 0

  fun isValid(): Boolean = beginIndex < endIndex && beginIndex >= 0

  fun clear() {
    beginIndex = 0
    endIndex = 0
  }

  fun offset(offset: Int) {
    if (isValid()) {
      beginIndex += offset
      endIndex += offset
    }
  }

  operator fun contains(index: Int): Boolean = index in beginIndex until endIndex
}

enum class AmountField {
  FixedFraction,
  ZeroNotation,
  CurrencySuffix,
}

class AmountFieldPositions internal constructor(
  val cursorPosition: Int,
  internal val fixedFraction: FieldRange,
  internal val zeroNotation: FieldRange,
) {
  companion object {
    val Empty: AmountFieldPositions = AmountFieldPositions(
      cursorPosition = -1,
      fixedFraction = FieldRange(),
      zeroNotation = FieldRange(),
    )
  }
}

File: camount-swift/Sources/Camount/Core/FieldRange.swift
import Foundation

struct FieldRange: Equatable {
    var beginIndex: Int
    var endIndex: Int

    init(beginIndex: Int = 0, endIndex: Int = 0) {
        self.beginIndex = beginIndex
        self.endIndex = endIndex
    }

    var length: Int { isValid ? endIndex - beginIndex : 0 }
    var isValid: Bool { beginIndex < endIndex && beginIndex >= 0 }

    mutating func clear() {
        beginIndex = 0
        endIndex = 0
    }

    mutating func offset(by value: Int) {
        guard isValid else { return }
        beginIndex += value
        endIndex += value
    }

    func contains(_ index: Int) -> Bool {
        index >= beginIndex && index < endIndex
    }
}

enum AmountFieldKind {
    case fixedFraction
    case zeroNotation
    case currencySuffix
}

struct AmountFieldPositions: Equatable {
    let cursorPosition: Int
    let fixedFraction: FieldRange
    let zeroNotation: FieldRange

    static let empty = AmountFieldPositions(
        cursorPosition: -1,
        fixedFraction: FieldRange(),
        zeroNotation: FieldRange()
    )
}
